lib/vector-core/src/event/mod.rs

```diff
@@ -43,6 +43,8 @@ pub const PARTIAL: &str = "_partial";
 
 #[derive(PartialEq, PartialOrd, Debug, Clone)]
 pub enum Event {
+    Chunk(Vec<u8>, EventMetadata),
+    Frame(Vec<u8>, EventMetadata),
```

Comment:
I'm pretty skeptical of this modeling. Chunks and frames are _not_ events, and I don't think we want to allow passing them all the way through every type of component in the topology. It seems to me that these concepts should be isolated to the components where they're useful/relevant. For example, would we really want to allow a source to output chunks or frames to a downstream transform?

Reply:
That's a good question. Are you skeptical of the terminology, the concept or both?

> I don't think we want to allow passing them all the way through every type of component in the topology.

I think that's a direct consequence of trying to make this concept as general as possible. Some inspiration for not distinguishing between framing and decoding/encoding have been [logstash's codecs](https://www.elastic.co/guide/en/logstash/current/codec-plugins.html). Trying to understand the argument more, do you have more specific thoughts on why that would feel wrong?

> For example, would we really want to allow a source to output chunks or frames to a downstream transform?

This would basically be some sort of configuration error if the user didn't provide enough codecs if the event leaves the source unstructured. But I could very well imagine that transforms implement the `Chunk`/`Frame` cases and coerce it to a type they can handle. On the upside, this would enable transforms on a very low level don't make any assumptions about the structured format.

Reply:
I'm skeptical of the concept. I don't think "as general as possible" should be the goal here. We're blurring the line between different layers of abstraction (framing is below parsing is below events) and essentially making a huge number of states representable that don't make much logical sense (e.g. what's a sink supposed to do if it receives a `Chunk`?). We could try to eliminate things like that with validation, but why put ourselves in that position in the first place when we could use the type system to make sure only expected arrangements are allowed?

More broadly, changing our event model like this dramatically increases the scope of this work and I don't see any benefit to that increased scope, just a lot of potential confusion. Why should framing work on events?

Reply:
> We're blurring the line between different layers of abstraction (framing is below parsing is below events)

I agree that blurring these lines should, if any, be a very deliberate choice. My first shot at this topic has also been to keep the framing/parsing/event separation, as outlined in the RFC https://github.com/timberio/vector/pull/7352. However, my opinion evolved from this first take as I moved further into the implementation, after having collected more context and running into some limitations of that model.

The current concept developed from a discussion I had with @jszwedko, where we came up with a few concrete cases and looked if they could be modeled with the separate framing / codec model.

One of these cases was "JSON log records, line-delimited, collectively gzipped and sent via UDP".

To ingest these logs, one would need to apply an
1. IO-delimited framer to separate a bundle of records at the UDP frame boundaries
2. gzip decompression / decoder
3. line-delimited framer
4. JSON decoder

In that case, we are moving up and down between the abstractions (1. framing -> 2. codec -> 3. framing -> 4. codec), which broke the previous assumption that you can't "go back" once you advance to the framing/parsing/event stage.

I hadn't anticipated something like that could happen, and would be very interested in how you would handle that case.

Reply:
@lukesteensen I rolled back to the initial approach using the separate framing and codec stages, for the sake of sticking to the original RFC and making progress here.

I still think that combining these in a single concept would allow the user to handle more inputs / outputs, regardless of in which order the input was compressed, delimited and encoded. However, this would better be discussed in a dedicated RFC as it would affect the overall architecture quite heavily.